File: updater/src/main/java/updater/identifiers/FloorDecoration.java
package updater.identifiers;

import org.objectweb.asm.Type;
import updater.utility.AbstractIdentifier;
import updater.utility.ClassWrapper;
import updater.utility.DependsOn;
import updater.utility.TypeUtilities;

@DependsOn({Entity.class, Scene.class})
public class FloorDecoration extends AbstractIdentifier {
    @Override
    public boolean isMatch(ClassWrapper classNode) {
        return classNode.isOwnerless() &&
                classNode.getCountFieldsOfType(TypeUtilities.getTypeOfIdentifiedClass("Entity")) == 1 &&
                classNode.getCountFieldsOfType(Type.INT_TYPE) == 4;
    }

    public FieldIdentifier entity = fieldIdentifier(fieldWrapper -> fieldWrapper.isOfType(TypeUtilities.getTypeOfIdentifiedClass("Entity")));

//    public FieldInMethodIdentifier x = fieldInMethodIdentifier(TypeUtilities.getTypeOfIdentifiedClass("Scene").getClassName(), TypeUtilities.getFieldOfIdentifiedClass("Scene","newFloorDecoration"));
}

File: find-names/src/main/java/mapper/identifiers/classes/FloorDecoration.kt
package mapper.identifiers.classes

import mapper.abstractclasses.IdentityMapper
import mapper.annotations.DependsOn
import mapper.identifiers.MethodPutField
import mapper.predicateutilities.and
import mapper.predicateutilities.predicateOf
import mapper.predicateutilities.type
import mapper.wrappers.ClassWrapper
import mapper.wrappers.FieldWrapper
import org.objectweb.asm.Type.INT_TYPE
import org.objectweb.asm.Type.LONG_TYPE

@DependsOn(Entity::class)
class FloorDecoration : IdentityMapper.Class() {
    override val predicate = predicateOf<ClassWrapper> { it.superType == Any::class.type }
            .and { it.instanceFields.count { it.type == type<Entity>() } == 1 }
            .and { it.instanceFields.count { it.type == INT_TYPE } == 4 }

    @DependsOn(Entity::class)
    class entity : InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == type<Entity>() }
    }

    @DependsOn(Scene.newFloorDecoration::class)
    class x : MethodPutField(Scene.newFloorDecoration::class, 0, INT_TYPE)

    @DependsOn(Scene.newFloorDecoration::class)
    class y : MethodPutField(Scene.newFloorDecoration::class, 1, INT_TYPE)

    @DependsOn(Scene.newFloorDecoration::class)
    class tileHeight : MethodPutField(Scene.newFloorDecoration::class, 2, INT_TYPE)

    @DependsOn(Scene.newFloorDecoration::class)
    class tag : MethodPutField(Scene.newFloorDecoration::class, 0, LONG_TYPE)

    @DependsOn(Scene.newFloorDecoration::class)
    class flags : MethodPutField(Scene.newFloorDecoration::class, 3, INT_TYPE)
}